crates/rethnet_eth/src/block/detailed.rs

```diff
@@ -126,3 +126,11 @@ impl Deref for DetailedBlock {
         &self.block
     }
 }
+
+impl PartialEq for DetailedBlock {
```

Comment:
Maybe I'm interpreting this too strictly, but from the [`PartialEq` docs](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html):

> This trait allows for partial equality, for types that do not have a full equivalence relation. For example, in floating point numbers NaN != NaN, so floating point types implement PartialEq but not [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html).

Since the cached hash can be in different states, I didn't derive the `Eq`, even though we know that they should logically be equal too.

Reply:
Thanks for clarifying! 

For me the absence of `Eq` when `PartialEq` is implemented is a warning sign that something weird is going with the type and I should be careful, so I find it helpful to implement whenever possible 🙏 .

I think we're ok to implement `Eq` as long as the compared members are `Eq` (this I haven't checked). We could be in trouble if we also compared the `hash` field as we could violate the reflexivity requirement  of [Eq:](https://doc.rust-lang.org/std/cmp/trait.Eq.html)

> - reflexive: a == a;

---

Regarding the `hash()` method, is it possible there exists two instance of `DetailedBlock`, let's call them `a` and `b`, such that `a != b /\ a.hash() == b.hash()`?

If so, I'd rename the `hash` method to `hash_block_header` as with a 256-bit hash space I'd expect that

`a.hash() == b.hash() => a == b`

Reply:
What does `/\` mean?

The hash is cryptographically accurate, so hash equality should imply block equality. People can construct a detailed block with the wrong transactions and addresses though, but since we check those, we should be guaranteed of equality

Reply:
> What does /\ mean?

Sorry, I use it as ASCII for the logical and symbol

> The hash is cryptographically accurate, so hash equality should imply block equality. People can construct a detailed block with the wrong transactions and addresses though, but since we check those, we should be guaranteed of equality

Ok if it's possible to construct two `DetailedBlock` instances such that they're not equal, but the `hash()` returns the same values, I'd rename the `hash` method to `hash_block_header` to make it clear to users that it doesn't hash the entire struct.

Reply:
Resolved in https://github.com/NomicFoundation/hardhat/pull/4300/commits/b490e3750b50b9d840e5a1be8c79d7c9f1a5c070

Reply:
Thanks, looks nice!